crates/curp/src/tracker.rs

```diff
@@ -0,0 +1,351 @@
+#![allow(unused)] // TODO remove when used
+
+use std::collections::VecDeque;
+use std::ops::{AddAssign, Sub};
+
+use clippy_utilities::NumericCast;
+
+/// Bits of usize
+const USIZE_BITS: usize = std::mem::size_of::<usize>() * 8;
+
+/// Default bit vec queue capacity, this is the number of inflight requests that a client expects.
+/// It can reduce about `log_2(DEFAULT_BIT_VEC_QUEUE_CAP)` `VecDequeue` memory reallocations.
+const DEFAULT_BIT_VEC_QUEUE_CAP: usize = 1024;
+
+/// A one-direction bit vector queue
+/// It use a ring buffer `VecDeque<usize>` to store bits
+///
+/// Memory Layout:
+///
+/// `010000000101_110111011111_000001000000000`
+///  ^        ^                ^    ^
+///  |        |________________|____|
+///  |________|         len    |____|
+///     head                    tail
```

Comment:
tail is a part of len?

Reply:
Tail indicates how many bits are used in the last usize.